### ISSUE TITLE
[.NET] Only the invariant culture works

### ISSUE THREAD

**Issue by rolfbjarne** [xamarin/xamarin-macios#8906]
### Steps to Reproduce

Launch any app.

Ref: https://github.com/dotnet/designs/pull/102#discussion_r422469705

### Expected Behavior

Successful launch

### Actual Behavior

<details>
<summary>warning: CLR: Managed code called FailFast, saying "Infinite recursion during resource lookup within System.Private.CoreLib.  This may be a bug in System.Private.CoreLib, or potentially in certain extensibility points such as assembly resolve events or CultureInfo names.  Resource name: Arg_DllNotFoundException"</summary>

```
2020-06-10 17:07:55.450901+0200 MySingleView[72595:4489235] warning: CLR: Managed code called FailFast, saying "Infinite recursion during resource lookup within System.Private.CoreLib.  This may be a bug in System.Private.CoreLib, or potentially in certain extensibility points such as assembly resolve events or CultureInfo names.  Resource name: Arg_DllNotFoundException"

=================================================================
        Native Crash Reporting
=================================================================
Got a abrt while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
        Native stacktrace:
=================================================================
        0x1064bef2b - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_dump_native_crash_info
        0x10646b441 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_handle_native_crash
        0x1064be782 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : sigabrt_signal_handler
        0x7fff51c005fd - /Applications/Xcode_11.5.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/system/libsystem_platform.dylib : _sigtramp
        0x0 - Unknown
        0x7fff51af0b7c - /Applications/Xcode_11.5.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/system/libsystem_c.dylib : abort
        0x10662f0a3 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : ves_icall_System_Environment_FailFast
        0x10662f0d7 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : ves_icall_System_Environment_FailFast_raw
        0x106b0a37f - Unknown
        0x106b0a2f3 - Unknown
        0x106b081bb - Unknown
        0x106b07ad3 - Unknown
        0x106b07a5b - Unknown
        0x106b079db - Unknown
        0x106b07813 - Unknown
        0x106b07961 - Unknown
        0x1063ee6f6 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_runtime_invoke
        0x10657815e - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : do_runtime_invoke
        0x106574de1 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_runtime_object_init_handle
        0x10652bc41 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_exception_from_token
        0x106b09c0b - Unknown
        0x106b09b33 - Unknown
        0x106b09a5b - Unknown
        0x106b099a3 - Unknown
        0x106b09963 - Unknown
        0x106b098f3 - Unknown
        0x106b098b3 - Unknown
        0x106b09863 - Unknown
        0x106b0939b - Unknown
        0x106b083dd - Unknown
        0x106b07ad3 - Unknown
        0x106b07a5b - Unknown
        0x106b079db - Unknown
        0x106b07813 - Unknown
        0x106b07961 - Unknown
        0x1063ee6f6 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_runtime_invoke
        0x10657815e - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : do_runtime_invoke
        0x106574de1 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_runtime_object_init_handle
        0x10652bc41 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_exception_from_token
        0x106b0775b - Unknown
        0x106b068eb - Unknown
        0x106b04d63 - Unknown
        0x106a590ce - Unknown
        0x1063ee6f6 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_runtime_invoke
        0x10657815e - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : do_runtime_invoke
        0x10657545b - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_runtime_class_init_full
        0x106414118 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_method_to_ir
        0x1063e4688 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mini_method_compile
        0x1063e6ce4 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_compile_method_inner
        0x1063ea4f7 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_compile_method_with_opt
        0x10646d5d8 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : common_call_trampoline
        0x10646d07d - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_magic_trampoline
        0x1069db396 - Unknown
        0x106b04a7b - Unknown
        0x106b02e83 - Unknown
        0x106b02d9b - Unknown
        0x106a590ce - Unknown
        0x1063ee6f6 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_runtime_invoke
        0x10657815e - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : do_runtime_invoke
        0x10657545b - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_runtime_class_init_full
        0x106414118 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_method_to_ir
        0x1063e4688 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mini_method_compile
        0x1063e6ce4 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_compile_method_inner
        0x1063ea4f7 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_compile_method_with_opt
        0x10646d5d8 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : common_call_trampoline
        0x10646d07d - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_magic_trampoline
        0x1069db396 - Unknown
        0x106b02d53 - Unknown
        0x106a590ce - Unknown
        0x1063ee6f6 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_runtime_invoke
        0x10657815e - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : do_runtime_invoke
        0x10657545b - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_runtime_class_init_full
        0x106414118 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_method_to_ir
        0x1063e4688 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mini_method_compile
        0x1063e6ce4 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_compile_method_inner
        0x1063ea4f7 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_jit_compile_method_with_opt
        0x10646d5d8 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : common_call_trampoline
        0x10646d07d - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libmonosgen-2.0.dylib : mono_magic_trampoline
        0x1069db396 - Unknown
        0x106afea83 - Unknown
        0x106afdc0b - Unknown
        0x106afd5eb - Unknown
        0x106afd223 - Unknown
        0x106afd04b - Unknown
        0x106afcf2b - Unknown
        0x106af98c3 - Unknown
        0x106aec7fb - Unknown
        0x106a646ab - Unknown
        0x106a64543 - Unknown
        0x106a644e3 - Unknown
        0x106a6433b - Unknown
        0x106a63f03 - Unknown
        0x106a63d93 - Unknown
        0x106a62b5b - Unknown
        0x10638dc47 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libxamarin.dylib : xamarin_register_assembly
        0x106390394 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libxamarin.dylib : xamarin_initialize
        0x1063984e0 - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/libxamarin.dylib : xamarin_main
        0x1062d4f5d - /Users/rolf/Library/Developer/CoreSimulator/Devices/289E372A-501C-4499-A1A6-59C5B3B6A9AE/data/Containers/Bundle/Application/62875EFF-27D1-40D9-87BD-F223F45E4EC6/MySingleView.app/MySingleView : main
        0x7fff51a231fd - /Applications/Xcode_11.5.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/system/libdyld.dylib : start
        0x1 - Unknown

=================================================================
        Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x7fff51b6133a):0x7fff51b6132a  ff ff c3 90 90 90 b8 48 01 00 02 49 89 ca 0f 05  .......H...I....
0x7fff51b6133a  73 08 48 89 c7 e9 e5 a2 ff ff c3 90 90 90 b8 53  s.H............S
0x7fff51b6134a  00 00 02 49 89 ca 0f 05 73 08 48 89 c7 e9 cd a2  ...I....s.H.....
0x7fff51b6135a  ff ff c3 90 90 90 b8 83 01 00 02 49 89 ca 0f 05  ...........I....

=================================================================
        Managed Stacktrace:
=================================================================
          at <unknown> <0xffffffff>
          at System.Environment:FailFast <0x0006e>
          at System.Environment:FailFast <0x00022>
          at System.SR:InternalGetResourceString <0x001ba>
          at System.SR:GetResourceString <0x00062>
          at System.SR:GetResourceString <0x0002a>
          at System.SR:get_Arg_DllNotFoundException <0x0001a>
          at System.DllNotFoundException:.ctor <0x00022>
          at System.Object:runtime_invoke_void__this__ <0x000e0>
          at <unknown> <0xffffffff>
          at Globalization:GetDefaultLocaleName <0x0005a>
          at System.Globalization.CultureData:GetDefaultLocaleName <0x00052>
          at System.Globalization.CultureInfo:GetUserDefaultCulture <0x0005a>
          at System.Globalization.CultureInfo:InitializeUserDefaultCulture <0x00022>
          at System.Globalization.CultureInfo:GetUserDefaultUICulture <0x00012>
          at System.Globalization.CultureInfo:InitializeUserDefaultUICulture <0x00022>
          at System.Globalization.CultureInfo:get_UserDefaultUICulture <0x00032>
          at System.Globalization.CultureInfo:get_CurrentUICulture <0x00052>
          at System.Resources.ResourceManager:GetString <0x0006a>
          at System.SR:InternalGetResourceString <0x003dc>
          at System.SR:GetResourceString <0x00062>
          at System.SR:GetResourceString <0x0002a>
          at System.SR:get_Arg_DllNotFoundException <0x0001a>
          at System.DllNotFoundException:.ctor <0x00022>
          at System.Object:runtime_invoke_void__this__ <0x000e0>
          at <unknown> <0xffffffff>
          at Globalization:LoadICU <0x0005a>
          at System.Globalization.GlobalizationMode:GetGlobalizationInvariantMode <0x0007a>
          at System.Globalization.GlobalizationMode:.cctor <0x00012>
          at System.Object:runtime_invoke_void <0x0007d>
          at <unknown> <0xffffffff>
          at System.Globalization.CultureData:CreateCultureWithInvariantData <0x00cda>
          at System.Globalization.CultureData:get_Invariant <0x00032>
          at System.Globalization.CultureInfo:.cctor <0x0001a>
          at System.Object:runtime_invoke_void <0x0007d>
          at <unknown> <0xffffffff>
          at System.Globalization.CompareInfo:.cctor <0x00012>
          at System.Object:runtime_invoke_void <0x0007d>
          at <unknown> <0xffffffff>
          at ObjCMethod:get_IsPropertyAccessor <0x000c2>
          at Registrar.Registrar:GetBindAsAttribute <0x0015a>
          at ObjCMethod:get_NativeReturnType <0x000ea>
          at Registrar.Registrar:ComputeSignature <0x001c2>
          at ObjCMethod:ComputeSignature <0x0005a>
          at ObjCMethod:ValidateSignature <0x0005a>
          at ObjCType:Add <0x000c2>
          at Registrar.Registrar:RegisterTypeUnsafe <0x047fa>
          at Registrar.Registrar:RegisterType <0x0008a>
          at Registrar.DynamicRegistrar:Register <0x00042>
          at ObjCRuntime.Class:Register <0x00032>
          at ObjCRuntime.Runtime:RegisterAssembly <0x000ea>
          at ObjCRuntime.Runtime:RegisterAssembly <0x00062>
          at ObjCRuntime.Runtime:register_assembly <0x00042>
          at ObjCRuntime.Runtime:register_assembly <0x0006a>
=================================================================
```

</details>

### Workaround

```c
setenv (""DOTNET_SYSTEM_GLOBALIZATION_INVARIANT"", ""1"", 1);
```

---

Time estimate: once the runtime ships support for it: 2 days on our side to verify other cultures work and update tests.

**Comment by filipnavara** [comment]
Seems like libSystem.Globalization.Native.dylib with the Mono runtime is not bundled properly, not even on macOS.

**Comment by rolfbjarne** [comment]
ICU support has landed: https://github.com/dotnet/runtime/pull/49406.

Bumping to latest .NET should get this (#10772), then we'll have to:

```
When you link in the libraries, you'll also need to link in c++, if you don't do that already.  You can enable ICU by getting rid of the invariant mode environment variable first and then set ICU_DAT_FILE_PATH to the icudt.dat file that is included in the runtime pack.  You should add that to the app's bundle.
ICU_DAT_FILE_PATH should be added to the appcontext keys / values when you call monovm_initialize at startup
then it should all "just work"
```

**Comment by spouliot** [comment]
> ICU support has landed: dotnet/runtime#49406.

Only for iOS.

IOW not for tvOS.

Situation about Catalyst apps remains unclear (whether or not Apple will accept the use of the system ICU).

**Comment by spouliot** [comment]
tvOS supported ICU (back ported to preview 4)

Catalyst does not (yet) and this is tracked with https://github.com/xamarin/xamarin-macios/issues/11392 so we can close this one
